l10n_nl_tax_statement/models/account_move_line.py

```diff
@@ -31,10 +31,52 @@ def _get_l10n_nl_vat_statement_protected_fields(self):
             "tax_tag_ids",
         ]
 
+    def check_field_is_equal(self, changed_protected_field, values):
+        self.ensure_one()
+        if self._fields.get(changed_protected_field) and self._fields[
+            changed_protected_field
+        ].type not in ["many2many", "one2many"]:
+            return self[changed_protected_field] == values[changed_protected_field]
+        # if field is X2M we must translate commands to values.
+        # the only acceptable value is [[6,0,self[changed_protected_field].ids]]
+        # wich is what the web client posts in case there is a editable X2M in form
+        # that is unchanged.
+        return (
+            values[changed_protected_field][0] == 6
+            and values[changed_protected_field][2] == self[changed_protected_field].ids
+        )
+
     def write(self, values):
-        if self._l10n_nl_vat_statement_should_check_write(values):
-            self._l10n_nl_vat_statement_check_state()
-        return super().write(values)
+        # before doing anything we check the nl_vat_statement:check_state,
+        # if this is not allowed it will raise and no further operations
+        # are necessary.
+        if not self._l10n_nl_vat_statement_should_check_write(values):
+            return super().write(values)
+        # now we add code to check if any of the modified fields present in
+        # values are the same as existing field value. this is a known limitation
+        # of odoo web client, it passes all non-readonly value fields in a form
+        # for writing , even if the value has not been changed.
+        res = True
+        protected_fields = self._get_l10n_nl_vat_statement_protected_fields()
+        protected_fields_in_values = [x for x in values.keys() if x in protected_fields]
+        for this in self:
+            # there are protected fields in dict, we proceed to check if such keys
+            # are in reality leaving the field the same.
+            # we need to recreate this dict every time
+            clean_values = dict(values)
+            for changed_protected_field in protected_fields_in_values:
+                is_equal = this.check_field_is_equal(
+                    changed_protected_field, clean_values
+                )
+                if is_equal:
+                    clean_values.pop(changed_protected_field)
+            # in case of recordset write we need to evaluate write one by one
+            # and re-check that our new clean dict is allowed
+            if this._l10n_nl_vat_statement_should_check_write(clean_values):
+                if this._l10n_nl_vat_statement_should_check_write(clean_values):
+                    this._l10n_nl_vat_statement_check_state()
+                res = res and super(AccountMoveLine, this).write(clean_values)
```

Comment:
You can have this out of (after) the loop, and still have the batched write: `return super().write(values)` .

Well, while you are at it, rename the argument `values` in the write method to `vals`  so it is the same as in the Odoo code.

Anyway in this way you do not need the res variable.

Reply:
No, the write needs to be done this way. We are checking for equality in values.  If we write on a recordset and are checking 

I have tried to do the complete write on "self" when possible, but in this specific case it is  not.

  I need to check the equality on every single record, "clean values" will change for each, and each one needs their separate check.

I could do what you say if  I pop out a protected value ONLY once i have verified that all  protected fields in dict are noop writes, that would work, but I would still need to check one-by one _l10n_nl_vat_statement_check_state  wich is not written for recordsets, but records.


I can remove the "res" though.

Reply:
Cannot do that.
THis checks for equality of protected fields record-per-record, i could do as you say by following the logic "if one record has a non noop protected write , make all fail", but i still need to  if 
               this._l10n_nl_vat_statement_should_check_write(clean_values):

for every record "clean dict" wlll be different and needs to be recreated, so i cannot just re-use it.

Example : 

         date is protected field:    I write 12th october 2023 on a recordset(1,2,3):

            on record 1 it is noop, value was already 12thoct 2023 --: ok
            on record 2 date value is DIFFERENT than 23-oct and state not draft ----> not okay.
            on record 2 date value is DIFFERENT than 23-oct and state not draft ----> not okay.
        
    "clean value" dict  would pop out "date" on rec 1, and then rec 2 and 3 would still be popped out.

   We could do it, but instead of raising error on REALLY Wrong records, it would silently fail sometimes and other times raise error depending on recordset  ordering.


We could implement this by doing:

                If the protected field is valid (equal value) in all records proceed, otherwise raise error, always.

Will do that.

Reply:
I am not suggesting to change anything in the checks, those should be done record by record, just to get the write out of the loop. As a failed check in any record will raise an exception, this will not make any functional difference: one failed records will prevent update of all records.

But with the batched write, this will be more efficient.

Reply:
done